### PR TITLE
[extractor/soundcloud] Implement user permalink support

### LINKED ISSUE #5841: Add download-by-userid support for SoundCloud
### DO NOT REMOVE OR SKIP THE ISSUE TEMPLATE

- [X] I understand that I will be **blocked** if I remove or skip any mandatory\* field

### Checklist

- [X] I'm requesting a site-specific feature
- [X] I've verified that I'm running yt-dlp version **2022.11.11** ([update instructions](https://github.com/yt-dlp/yt-dlp#update)) or later (specify commit)
- [X] I've checked that all provided URLs are playable in a browser with the same IP and same login details
- [X] I've searched the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar issues **including closed ones**. DO NOT post duplicates
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)
- [ ] I've read about [sharing account credentials](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#are-you-willing-to-share-account-details-if-needed) and I'm willing to share it if required

### Region

United States

### Example URLs

https://soundcloud.com/neilcic
https://w.soundcloud.com/player/?url=https%3A%2F%2Fapi.soundcloud.com%2Fusers%2F30909869 (correspond to the above profile URL)

### Provide a description that is worded well enough to be understood

I would like the ability to use a more persistent identifier to download tracks from SoundCloud users.  The profile URL can be changed at any time, and as far as I'm aware yt-dlp currently does not support downloads through a user's userid (or `uploader_id` as it's rendered when passing `-j`).

SoundCloud doesn't expose a permanent URL from the userid so I'm not sure how such a thing would be passed to yt-dlp, but a number of `meta` elements on the page use a protocol-like string in the form `soundcloud://users:30909869`.

**Edit:** yt-dlp already supports track permalinks in the form `https://api.soundcloud.com/tracks/902235112`, so `https://api.soundcloud.com/users/30909869` for user permalinks would suffice.

No account is required to access the example URLs above.

### Provide verbose output that clearly demonstrates the problem

- [X] Run **your** yt-dlp command with **-vU** flag added (`yt-dlp -vU <your command line>`)
- [X] Copy the WHOLE output (starting with `[debug] Command-line config`) and insert it below

### Complete Verbose Output

```shell
[debug] Command-line config: ['-vU', 'https://w.soundcloud.com/player/?url=https%3A%2F%2Fapi.soundcloud.com%2Fusers%2F30909869']
[debug] Encodings: locale cp1252, fs utf-8, pref cp1252, out utf-8, error utf-8, screen utf-8
[debug] yt-dlp version 2022.11.11 [8b64402] (pip)
[debug] Python 3.9.2 (CPython AMD64 64bit) - Windows-10-10.0.19041-SP0 (OpenSSL 1.1.1i  8 Dec 2020)
[debug] exe versions: ffmpeg 5.0.1-full_build-www.gyan.dev (setts), ffprobe 5.0.1-full_build-www.gyan.dev
[debug] Optional libraries: Cryptodome-3.10.1, brotli-1.0.9, certifi-2020.12.05, mutagen-1.45.1, sqlite3-2.6.0, websockets-10.0
[debug] Proxy map: {}
[debug] Loaded 1723 extractors
[debug] Fetching release info: https://api.github.com/repos/yt-dlp/yt-dlp/releases/latest
Latest version: 2022.11.11, Current version: 2022.11.11
yt-dlp is up to date (2022.11.11)
[debug] [SoundcloudEmbed] Extracting URL: https://w.soundcloud.com/player/?url=https%3A%2F%2Fapi.soundcloud.com%2Fusers%2F30909869
[debug] [generic] Extracting URL: https://api.soundcloud.com/users/30909869
[generic] 30909869: Downloading webpage
ERROR: [generic] Unable to download webpage: HTTP Error 401: Unauthorized (caused by <HTTPError 401: 'Unauthorized'>); please report this issue on  https://github.com/yt-dlp/yt-dlp/issues?q= , filling out the appropriate issue template. Confirm you are on the latest version using  yt-dlp -U
```

### PR DESCRIPTION
**Pull request by nosoop**
**IMPORTANT**: PRs without the template will be CLOSED

### Description of your *pull request* and other information

</details>

Fixes #5841.

Adds support for downloading from SoundCloud users via ID, which is resistant to profile URL changes.  URLs are in the form `https://api.soundcloud.com/users/30909869`.

~~This PR also adds `_API_V1_BASE`, as the resolve endpoint does not work with the V2 subdomain.~~

Only the 'all' resource is supported.

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Fix or improvement to an extractor (Make sure to add/update tests)
- [ ] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

### CODE REVIEW

File: yt_dlp/extractor/soundcloud.py [hunk 1]
```diff
@@ -769,12 +778,17 @@ class SoundcloudUserIE(SoundcloudPagedPlaylistBaseIE):
...
 
-        user = self._download_json(
-            self._resolv_url(self._BASE_URL + uploader),
-            uploader, 'Downloading user info', headers=self._HEADERS)
+        if user_id:
+            user_info_url = self._resolv_url(self._API_V1_BASE + 'users/' + user_id)
+            resource = 'all'
+        else:
+            user_info_url = self._resolv_url(self._BASE_URL + uploader)
+            resource = mobj.group('rsrc') or 'all'
```
Comment (pukkandan) [inline yt_dlp/extractor/soundcloud.py:788]: Since it's a completely different URL pattern and code path, better to create a new class

**Comment by nosoop** [comment]
Moved to separate class as requested.  The expression no longer matches against `api-v2`, since that was never a valid form for the permalink.

Was considering skipping downloading the user info, but realized other information extractors still do so.

File: yt_dlp/extractor/soundcloud.py [hunk 2]
```diff
@@ -782,6 +782,31 @@ def _real_extract(self, url):
...
+        mobj = self._match_valid_url(url)
+        user_id = mobj.group('user_id')
+
+        user = self._download_json(
+            self._resolv_url(url), 'user:' + user_id,
+            'Downloading user info', headers=self._HEADERS)
+
+        return self._extract_playlist(
+            self._API_V2_BASE + 'stream/users/%s' % user['id'],
+            str_or_none(user.get('id')), '%s (permalink)' % user['username'])
```
Comment (pukkandan) [inline yt_dlp/extractor/soundcloud.py:807]:
```suggestion
class SoundcloudUserPermalinkIE(SoundcloudPagedPlaylistBaseIE):
    _VALID_URL = r'https?://api\.soundcloud\.com/users/(?P<id>\d+)'
    IE_NAME = 'soundcloud:user:permalink'
    _TESTS = [{
        'url': 'https://api.soundcloud.com/users/30909869',
        'info_dict': {
            'id': '30909869',
            'title': 'neilcic',
        },
        'playlist_mincount': 23,
    }]

    def _real_extract(self, url):
        user_id = self._match_id(url)
        user = self._download_json(
            self._resolv_url(url), user_id, 'Downloading user info', headers=self._HEADERS)

        return self._extract_playlist(
            f'{self._API_V2_BASE}stream/users/{user["id"]}',
            str_or_none(user.get('id')), user.get('username'))
```

File: yt_dlp/extractor/soundcloud.py [hunk 3]
```diff
@@ -782,6 +782,28 @@ def _real_extract(self, url):
...
+    }]
+
+    def _real_extract(self, url):
+        user_id = self._match_id(url)
+        user = self._download_json(
+            self._resolv_url(url), user_id, 'Downloading user info', headers=self._HEADERS)
+
+        return self._extract_playlist(
+            f'{self._API_V2_BASE}stream/users/{user["id"]}',
+            str_or_none(user.get('id')), user.get('username'))
```
Comment (pukkandan) [inline yt_dlp/extractor/soundcloud.py:804]:
```suggestion
            f'{self._API_V2_BASE}stream/users/{user["id"]}', user['id'], user.get('username'))
```

**Comment by pukkandan** [comment]
pls verify my changes

**Comment by nosoop** [comment]
Had to stringify `user['id']` for the test to pass but otherwise LGTM.

**Pull request merged**